### PR TITLE
refactoring the code to reduce all the boilerplate 

### PR DESCRIPTION
**Pull request by AlexMontgomerie**

### CODE REVIEW

**Comment by AlexMontgomerie** [comment]
One of the main changes from the frontend side is that the class constructors only accept keywords.

File: tests/test_layers.py [hunk 1]
```diff
@@ -1,8 +1,14 @@
 import unittest
 import ddt
 import json
+import sys
+
+sys.path.append("/home/alex/fpgaconvnet-model-rebase/")
```
Comment (ptoupas) [inline tests/test_layers.py:6]: I guess this one was left here not in purpose. We should probably remove that.

File: tests/test_layers.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
+sys.path.append("../fpgaconvnet/")
```
Comment (ptoupas) [inline tests/test_layers.py:7]: I kind of dont like to explicitly add things to the python path this way.
I would suggest to find a better way to import what is needed from the fpgaconvnet folder

Comment (AlexMontgomerie) [reply tests/test_layers.py:7]: Yeah it was on accident, I'll remove it.

File: tests/test_layers.py (hunk 1 above)
Comment (AlexMontgomerie) [reply tests/test_layers.py:6]: removed in  commit 4ccbb5749b9d18e883ea8b6a46e9d34d1d9bffe4

**Comment by AlexMontgomerie** [comment]
I've tried out the registry idea in 15b2828, and really like it ! Here's where it's instantiated for the Layer:

https://github.com/AlexMontgomerie/fpgaconvnet-model/blob/15b282894e45ae4312fcb1ab3ed494e4cd1f5879/fpgaconvnet/models/layers/Layer.py#L22C1-L41C20

It means that you can simply instantiate any class with 
```
Layer.build_from_dict("ConvolutionLayer", config)
```

File: tests/test_layers.py [hunk 3]
```diff
@@ -59,11 +69,12 @@ def run_test_wait_depth(self,layer):
     def run_test_resources(self,layer):
         # check resources
         rsc = layer.resource()
-        self.assertEqual(set(list(rsc.keys())), set(["BRAM","DSP","LUT","FF"]))
+        # self.assertEqual(set(list(rsc.keys())), set(["URAM", "BRAM","DSP","LUT","FF"]))
         self.assertTrue(rsc["LUT"] >= 0)
         self.assertTrue(rsc["FF"] >= 0)
         self.assertTrue(rsc["DSP"] >= 0)
         self.assertTrue(rsc["BRAM"] >= 0)
+        # self.assertTrue(rsc["URAM"] >= 0)
```
Comment (ptoupas) [inline tests/test_layers.py:77]: We could maybe enable or disable this line based on the fpga that is loaded for a specific run/experiment and its support of URAM or not

**Review by ptoupas — APPROVED** [review]

File: fpgaconvnet/models/layers/Layer.py [hunk 4]
```diff
@@ -66,283 +55,181 @@ def __post_init__(self):
...
     def get_coarse_out_feasible(self):
-        return get_factors(int(self.channels_out()))
+        pass
 
+    @abstractmethod
     def update(self):
         pass
 
     def layer_info(self, parameters, batch_size=1):
-        parameters.batch_size   = batch_size
```
Comment (Yu-Zhewen) [inline fpgaconvnet/models/layers/Layer.py:332]: why are we dropping some fields?

File: fpgaconvnet/models/layers/ConvolutionLayer.py [hunk 5]
```diff
@@ -106,13 +75,12 @@ def __init__(
             self.use_uram = False
         elif self.backend == "chisel":
             self.double_buffered = False
-            self.stream_weights = stream_weights
```
Comment (Yu-Zhewen) [inline fpgaconvnet/models/layers/ConvolutionLayer.py:109]: I'm not sure why this is deleted?

File: fpgaconvnet/models/layers/ConvolutionLayer3D.py [hunk 6]
```diff
@@ -20,93 +21,70 @@
...
         # weights buffering flag
         if self.backend == "hls":
             self.double_buffered = False
-            self.stream_weights = False
+            self.stream_weights = 0
             self.data_packing = False
+            self.use_uram = False
         elif self.backend == "chisel":
             self.double_buffered = False
-            self.stream_weights = False
```
Comment (Yu-Zhewen) [inline fpgaconvnet/models/layers/ConvolutionLayer3D.py:104]: same here

File: fpgaconvnet/models/layers/InnerProductLayer.py [hunk 7]
```diff
@@ -74,10 +50,6 @@ def __init__(
             self.data_packing = True
```
Comment (Yu-Zhewen) [inline fpgaconvnet/models/layers/InnerProductLayer.py:50]: self.stream_weights = stream_weights

File: fpgaconvnet/models/layers/InnerProductLayer3D.py [hunk 8]
```diff
@@ -13,59 +14,46 @@
...
         # weights buffering flag
         if self.backend == "hls":
             self.double_buffered = False
-            self.stream_weights = False
+            self.stream_weights = 0
             self.data_packing = False
+            self.use_uram = False
         elif self.backend == "chisel":
             self.double_buffered = False
             self.stream_weights = False
```
Comment (Yu-Zhewen) [inline fpgaconvnet/models/layers/InnerProductLayer3D.py:54]: same here

File: tests/test_layers.py (hunk 3 above)
Comment (ptoupas) [reply tests/test_layers.py:77]: @AlexMontgomerie Is there any update here? If not lets resolve the conversation

File: fpgaconvnet/models/layers/InnerProductLayer.py [hunk 9]
```diff
@@ -95,21 +67,36 @@ def __init__(
...
+            case "coarse_group":
+                assert(value in self.get_coarse_group_feasible())
+                super().__setattr__(name, value)
+                self.update()
 
-    @filters.setter
-    def filters(self, val: int) -> None:
-        self._filters = val
-        # self.update()
+            case "fine":
```
[the remark this answers is not in the document]
Comment (ptoupas) [reply fpgaconvnet/models/layers/InnerProductLayer.py:90]: @AlexMontgomerie Is there any update here? If not lets resolve the conversation

File: fpgaconvnet/models/layers/Layer.py (hunk 4 above)
Comment (ptoupas) [reply fpgaconvnet/models/layers/Layer.py:332]: @AlexMontgomerie I guess these are inherited from the base LayeBase class, but is this true @AlexMontgomerie or we need to add them back?

File: fpgaconvnet/models/layers/ConvolutionLayer.py (hunk 5 above)
Comment (ptoupas) [reply fpgaconvnet/models/layers/ConvolutionLayer.py:109]: @AlexMontgomerie Any idea about this one?

File: fpgaconvnet/models/layers/ConvolutionLayer3D.py (hunk 6 above)
Comment (ptoupas) [reply fpgaconvnet/models/layers/ConvolutionLayer3D.py:104]: @AlexMontgomerie Any idea about this one?

**Review by ptoupas — COMMENTED** [review]
Hey @Yu-Zhewen @AlexMontgomerie I guess its been a while this is an open PR.
Do you think we should start thinking about merging this one or its too early and you want to work on the module's refactoring first and and them together into `dev`?

**Comment by Yu-Zhewen** [comment]
I think it's better we close this now, and then we can deal with module's refactoring in a separate PR?

**Comment by ptoupas** [comment]
Yeah I also agree that we need to close this one first.
Even though I am not really sure about how functional everything is...

**Pull request closed without merging**